### PR TITLE
basic toolchain integration

### LINKED ISSUE #68: Assertion failed: (TablesInitialized && "backend forgot to call computeTables")
**Describe the bug**
When wompiling a C file with a void main function only containing a variable declaration, the pipeline fails.

**To Reproduce**
```c
void main(void) {
    // int asdf = 1234;
}
```
```sh
clang void.c
```

**Expected behavior**
The invocation of the clang should produce an ELF file (a.out)

**Screenshots**
test-inputs/void.c:4:16: warning: implicit conversion from 'long' to 'int' changes value from 1234 to -46 [-Wconstant-conversion]
    4 |     int asdf = 1234;
      |         ~~~~   ^~~~
Assertion failed: (TablesInitialized && "backend forgot to call computeTables"), function getAspectAction, file LegacyLegalizerInfo.cpp, line 198.
lowerFormallowerReturnPLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/bin/clang-17 -cc1 -triple picmid -emit-obj -mrelax-all -dumpdir a- -disable-free -clear-ast-before-backend -main-file-name void.c -mrelocation-model static -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -ffreestanding -std=c17 -mframe-pointer=none -debugger-tuning=gdb -target-linker-version 1053.12 -fcoverage-compilation-dir=/Users/hannes/Desktop/Studienarbeit/llvm-pic-experiments -resource-dir /Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/clang/17 -I/opt/homebrew/include/opencv4/ -I/opt/homebrew/include/opencv4/ -I. -cxx-isystem /opt/homebrew/include -cxx-isystem /opt/homebrew/include -cxx-isystem . -internal-isystem /Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/clang/17/include -fdebug-compilation-dir=/Users/hannes/Desktop/Studienarbeit/llvm-pic-experiments -ferror-limit 19 -fgnuc-version=4.2.1 -fcolor-diagnostics -faddrsig -o /var/folders/nf/sskb9mcn3vs15bbr3f8q2rx80000gn/T/void-b38e78.o -x c test-inputs/void.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'test-inputs/void.c'.
4.      Running pass 'Legalizer' on function '@main'
 #0 0x0000000111370f04 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMSupport.dylib+0x290f04)
 #1 0x00000001113714a0 PrintStackTraceSignalHandler(void*) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMSupport.dylib+0x2914a0)
 #2 0x000000011136f188 llvm::sys::RunSignalHandlers() (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMSupport.dylib+0x28f188)
 #3 0x00000001113726f0 SignalHandler(int) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMSupport.dylib+0x2926f0)
 #4 0x00000001966b3584 (/usr/lib/system/libsystem_platform.dylib+0x180477584)
 #5 0x0000000196682c20 (/usr/lib/system/libsystem_pthread.dylib+0x180446c20)
 #6 0x000000019658fa20 (/usr/lib/system/libsystem_c.dylib+0x180353a20)
 #7 0x000000019658ed10 (/usr/lib/system/libsystem_c.dylib+0x180352d10)
 #8 0x000000011222c3ac llvm::LegacyLegalizerInfo::getAspectAction(llvm::InstrAspect const&) const (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMGlobalISel.dylib+0x2303ac)
 #9 0x000000011222db30 llvm::LegacyLegalizerInfo::getAction(llvm::LegalityQuery const&) const (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMGlobalISel.dylib+0x231b30)
#10 0x0000000112223f88 llvm::LegalizerInfo::getAction(llvm::LegalityQuery const&) const (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMGlobalISel.dylib+0x227f88)
#11 0x000000011222423c llvm::LegalizerInfo::getAction(llvm::MachineInstr const&, llvm::MachineRegisterInfo const&) const (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMGlobalISel.dylib+0x22823c)
#12 0x00000001121d7f94 llvm::LegalizerHelper::legalizeInstrStep(llvm::MachineInstr&, llvm::LostDebugLocObserver&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMGlobalISel.dylib+0x1dbf94)
#13 0x00000001121c2f50 llvm::Legalizer::legalizeMachineFunction(llvm::MachineFunction&, llvm::LegalizerInfo const&, llvm::ArrayRef<llvm::GISelChangeObserver*>, llvm::LostDebugLocObserver&, llvm::MachineIRBuilder&, llvm::GISelKnownBits*) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMGlobalISel.dylib+0x1c6f50)
#14 0x00000001121c4a80 llvm::Legalizer::runOnMachineFunction(llvm::MachineFunction&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMGlobalISel.dylib+0x1c8a80)
#15 0x0000000109d93e04 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMCodeGen.dylib+0x543e04)
#16 0x000000011479686c llvm::FPPassManager::runOnFunction(llvm::Function&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMCore.dylib+0x3ae86c)
#17 0x000000011479dcf4 llvm::FPPassManager::runOnModule(llvm::Module&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMCore.dylib+0x3b5cf4)
#18 0x00000001147970f0 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMCore.dylib+0x3af0f0)
#19 0x0000000114796c74 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMCore.dylib+0x3aec74)
#20 0x000000011479e0fc llvm::legacy::PassManager::run(llvm::Module&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libLLVMCore.dylib+0x3b60fc)
#21 0x0000000105287dd4 (anonymous namespace)::EmitAssemblyHelper::RunCodegenPipeline(clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream>>&, std::__1::unique_ptr<llvm::ToolOutputFile, std::__1::default_delete<llvm::ToolOutputFile>>&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangCodeGen.dylib+0x3fdd4)
#22 0x000000010526aca4 (anonymous namespace)::EmitAssemblyHelper::EmitAssembly(clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream>>) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangCodeGen.dylib+0x22ca4)
#23 0x0000000105269d7c clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream>>) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangCodeGen.dylib+0x21d7c)
#24 0x0000000105bfddd0 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangCodeGen.dylib+0x9b5dd0)
#25 0x000000011a65e9a0 clang::ParseAST(clang::Sema&, bool, bool) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangParse.dylib+0x69a0)
#26 0x000000010c0901a8 clang::ASTFrontendAction::ExecuteAction() (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangFrontend.dylib+0x3241a8)
#27 0x0000000105bfbe34 clang::CodeGenAction::ExecuteAction() (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangCodeGen.dylib+0x9b3e34)
#28 0x000000010c08f970 clang::FrontendAction::Execute() (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangFrontend.dylib+0x323970)
#29 0x000000010bf8d9f0 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangFrontend.dylib+0x2219f0)
#30 0x00000001030daccc clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/lib/libclangFrontendTool.dylib+0x6ccc)
#31 0x00000001029c40f4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/bin/clang-17+0x10001c0f4)
#32 0x00000001029b27bc ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/bin/clang-17+0x10000a7bc)
#33 0x00000001029b14c8 clang_main(int, char**, llvm::ToolContext const&) (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/bin/clang-17+0x1000094c8)
#34 0x00000001029ebbb0 main (/Users/hannes/Desktop/Studienarbeit/llvm-pic/llvm/build/bin/clang-17+0x100043bb0)
#35 0x00000001962fa0e0 
clang: error: unable to execute command: Abort trap: 6
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 17.0.6 (https://github.com/llvm-pic/llvm-pic.git 799501a43e01df2471edc785429aefa73f932604)
Target: picmid
Thread model: posix
InstalledDir: /Users/hannes/Desktop/Studienarbeit/llvm-pic-experiments/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/nf/sskb9mcn3vs15bbr3f8q2rx80000gn/T/void-ccceee.c
clang: note: diagnostic msg: /var/folders/nf/sskb9mcn3vs15bbr3f8q2rx80000gn/T/void-ccceee.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /Users/hannes/Library/Logs/DiagnosticReports/clang-17_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg: 

********************

**Desktop (please complete the following information):**
 - OS: MacOS 
- Version 14.4.1

**Additional context**
test cases are not yet in the repo

https://github.com/llvm-pic/llvm-pic/commit/d12067206e6253d3ce53615f771432a8e75c825d

### PR DESCRIPTION
**Pull request by handiener**
fixes https://github.com/llvm-pic/llvm-pic/issues/68

merged working-ir branch with linker-startup branch to assemble the toolchain for the first time

also made some changes to the toolchain, mainly with the arguments passed to the linker

It is now possible to compile an empty main function. For that a regression test has been created: https://github.com/llvm-pic/llvm-pic/blob/integr-tst/clang/test/Driver/picmid-empty-void-main.c
Also further tests with linker and startup files have now documented in the new (private because of maybe microchip rights) experiments repository: https://github.com/llvm-pic/experiments

### CODE REVIEW

File: clang/lib/Driver/ToolChains/PICMid.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+using namespace clang::driver;
+using namespace clang::driver::tools;
+using namespace clang::driver::toolchains;
+
+PICMidToolChain::PICMidToolChain(const Driver &D, const llvm::Triple &Triple,
+         const llvm::opt::ArgList &Args)
+    : ToolChain(D, Triple, Args) {
+  // Look for binaries in both the installation and driver directory.
+  getProgramPaths().push_back(getDriver().getInstalledDir());
+  if (getDriver().getInstalledDir() != getDriver().Dir)
```
Comment (pr1metine) [inline clang/lib/Driver/ToolChains/PICMid.cpp:28]: Wrap "if body" with squirelly brackets `{}` pls.

File: clang/lib/Driver/ToolChains/PICMid.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+void PICMidToolChain::addClangTargetOptions(const ArgList &DriverArgs,
+                                ArgStringList &CC1Args,
+                                Action::OffloadKind) const {
+  CC1Args.push_back("-nobuiltininc");
+  // set to freestanding environment (main return type void)
+  CC1Args.push_back("-ffreestanding");
+  // set language Standard to C17
+  CC1Args.push_back("-std=c17");
+  // omit frame pointers because we don't use frame pointers for our simulated stack
+  CC1Args.push_back("-mframe-pointer=none");
```
Comment (pr1metine) [inline clang/lib/Driver/ToolChains/PICMid.cpp]: It's fine but I'm curious: Does Clang still produce `G_FRAME_INDEX` gMIR?

File: clang/lib/Headers/picmidintrin.h
Comment (pr1metine) [inline clang/lib/Headers/picmidintrin.h:1]: Is this file just a placeholder? Is this file included somewhere?

File: clang/test/Driver/picmid-empty-void-main.c [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (pr1metine) [inline clang/test/Driver/picmid-empty-void-main.c]: Remove extra line pls

File: llvm/lib/Target/PICMid/Examples/main.c
Comment (pr1metine) [inline llvm/lib/Target/PICMid/Examples/main.c:1]: Can you delete this example folder pls? We now have your experiments repo.

File: llvm/lib/Target/PICMid/GISel/PICMidLegalizerInfo.cpp [hunk 4]
```diff
@@ -89,3 +135,97 @@ PICMidLegalizerInfo::PICMidLegalizerInfo() {
 
   // Miscellaneous
```
Comment (pr1metine) [inline llvm/lib/Target/PICMid/GISel/PICMidLegalizerInfo.cpp:136]: You need to compute the Legalizer table, right? If so, pls include the fix as outlined here: https://github.com/llvm-pic/llvm-pic/issues/68#issuecomment-2104545572.

File: main.s
Comment (pr1metine) [inline main.s:1]: Pls delete

File: llvm/utils/lit/lit/llvm/config.py [hunk 5]
```diff
@@ -331,6 +331,8 @@ def get_clang_has_lsan(self, clang, triple):
         return False
 
     def make_itanium_abi_triple(self, triple):
+        if triple == "picmid":
```
Comment (pr1metine) [inline llvm/utils/lit/lit/llvm/config.py]: As discussed, you can define a new triple.

File: llvm/utils/lit/lit/llvm/config.py [hunk 6]
```diff
@@ -342,6 +344,8 @@ def make_itanium_abi_triple(self, triple):
         return m.group(1) + "-" + m.group(2) + "-" + m.group(3) + "-gnu"
 
     def make_msabi_triple(self, triple):
+        if triple == "picmid":
```
Comment (pr1metine) [inline llvm/utils/lit/lit/llvm/config.py]: As discussed, you can define a new triple.

**Review by pr1metine — CHANGES_REQUESTED** [review]
Ok, good work. Pls say "fixes #68" in your PR description. This links your PR to this issue. It is automatically closed when your PR gets merged. Also, if I recall correctly we did not just cherry pick Lennart's `working-ir` branch. We merged all of it into your branch. Please update your description accordingly if this is true.

**Comment by pr1metine** [comment]
And could you also extend this CI file to run `check-clang-drivers` on every PR push?

https://github.com/llvm-pic/llvm-pic/blob/d88bf266b492ff7182a84d3907b871f050a21955/.github/workflows/picmid-llvm-tests.yml#L30-L54

File: clang/lib/Driver/ToolChains/PICMid.cpp (hunk 1 above)
Comment (handiener) [reply clang/lib/Driver/ToolChains/PICMid.cpp:28]: your are right with the merging

File: clang/lib/Driver/ToolChains/PICMid.cpp (hunk 2 above)
Comment (handiener) [reply clang/lib/Driver/ToolChains/PICMid.cpp]: I am acutally unsure, maybe we can check that together. Do we need `G_FRAME_INDEX`?

File: clang/lib/Headers/picmidintrin.h
Comment (handiener) [reply clang/lib/Headers/picmidintrin.h:1]: It's just a placeholder. Should it be excluded from the merger?

File: llvm/lib/Target/PICMid/Examples/main.c
Comment (handiener) [reply llvm/lib/Target/PICMid/Examples/main.c:1]: Shure. This was there before though.

File: clang/lib/Driver/ToolChains/PICMid.cpp (hunk 2 above)
Comment (handiener) [reply clang/lib/Driver/ToolChains/PICMid.cpp]: agreed to leave it out

File: clang/lib/Headers/picmidintrin.h
Comment (handiener) [reply clang/lib/Headers/picmidintrin.h:1]: agreed to exclude

**Comment by pr1metine** [comment]
Don't quote me on that but you probably also need to add `LLVM_DEFAULT_TARGET_TRIPLE=<our-triple-name>` to the CMake args in the following line:

https://github.com/llvm-pic/llvm-pic/blob/d88bf266b492ff7182a84d3907b871f050a21955/.github/workflows/picmid-llvm-project-tests.yml#L92

That should fix your Clang CI problems.

**Comment by pr1metine** [comment]
Nice

**Review by pr1metine — APPROVED** [review]

**Comment by pr1metine** [comment]
Be sure to squash commit this

**Review by indicatelovelace — APPROVED** [review]

**Pull request merged**